File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.py
#Brute force solution of the problem
# class Solution:
#     def equalPairs(self, grid: List[List[int]]) -> int:
#         count = 0
#         n = len(grid)
#         for r in range(n):
#             for c in range(n):
#                 is_equal = True
#                 for i in range(n):
#                     if grid[r][i] != grid[i][c]:
#                         is_equal = False
#                 count += is_equal
#         return count

#Better approach
# class Solution:
#     def equalPairs(self, grid: List[List[int]]) -> int:
#         BASE = 131
#         MOD = 1_000_000_007
#         n = len(grid)
#         row_hashes = [0] * n
#         col_hashes = [0] * n
#         for i in range(n):
#             RowH = 0
#             ColH = 0
#             for j in range(n):
#                 RowH = (RowH * BASE + grid[i][j]) % MOD
#                 ColH = (ColH * BASE + grid[j][i]) % MOD
#             row_hashes[i] = RowH
#             col_hashes[i] = ColH
#         count = 0
#         for k in range(n):
#             for m in range(n):
#                 if row_hashes[k] == col_hashes[m]:
#                     count += 1
#         return count

class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        row_counts = Counter(tuple(row) for row in grid)
        return sum(row_counts[col] for col in zip(*grid))

File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.c
//Brute force solution of this problem 
// int equalPairs(int** grid, int gridSize, int* gridColSize) {
//     int count = 0;
//     for(int r = 0; r < gridSize; r++){
//         for(int c = 0; c < gridSize; c++){
//             int is_equal = true;
//             for(int i = 0; i < gridSize; i++){
//                 if(grid[r][i] != grid[i][c]){
//                     is_equal = false;
//                 }
//             }
//             if(is_equal){
//                 count++;
//             }
//         }
//     }
//     return count;
// }

#define BASE 131
#define MOD 1000000007

int equalPairs(int** grid, int gridSize, int* gridColSize){
    int RowHashes[gridSize];
    int ColHashes[gridSize];
    for(int i = 0; i < gridSize; i++){
        long long RowH = 0;
        long long ColH = 0;
        for(int j = 0; j < gridSize; j++){
            RowH = (RowH * BASE + grid[i][j]) % MOD;
            ColH = (ColH * BASE + grid[j][i]) % MOD;
        }
        RowHashes[i] = RowH;
        ColHashes[i] = ColH;
    }
    int count = 0;
    for(int i = 0; i < gridSize; i++){
        for(int j = 0; j < gridSize; j++){
            if(RowHashes[i] == ColHashes[j]){
                count++;
            }
        }
    }

    return count;
}